### ISSUE TITLE
RFC: Http interceptors and transformers

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#2684]
This issue is intended to define requirements for how the `Http` library should support common operations, like request and response transformations and error handling. The issue doesn't yet have a proposed design, but discussion, comments and suggestions are welcomed. Un-captured use cases are especially welcomed.

# History

Two widely-used feature of angular1 are [request and response interceptors](https://docs.angularjs.org/api/ng/service/$http#interceptors) and [request/response transformers](https://docs.angularjs.org/api/ng/service/$http#interceptors). Interceptors and Transformers have different goals and constraints.

 * Interceptors are expected to have side effects, while transformers' exclusive purpose is to mutate requests or responses synchronously.
 * Interceptors may be asynchronous, tranformers must be synchronous
 * Interceptors are global and configured in Angular 1's "compile" phase, while transformers may be global or local. Global transformations can be mutated in the "run" phase.
 * Error interceptors are declared explicitly and separately from normal interceptors, whereas transforms handle normal and error cases.
 * Error interceptors can recover from errors by returning a resolved promise

Some common use cases for interceptors:

 * Setting timeouts for requests
 * Setting custom XSRF tokens to request headers, as described in [$http docs](https://docs.angularjs.org/api/ng/service/$http#cross-site-request-forgery-xsrf-protection)
 * Add conditional parameters to all request urls (like an access token, or a/b-testing info)
 * Check authentication state before performing request, prompt user to authenticate
 * Have side effects on other parts of application (i.e. if response says user needs to authenticate, route them to login page)
 * Log all response errors returned from the server to a persisted log
 * Log total request/response time in order to do performance analysis
 * Filter out un-interesting request/responses from logs
 * Recover from request exceptions, or response errors (i.e. if request times out, return old data from cache)
 * Unwrapping responses (i.e. if response is {data: {foo:bar}}, return response.data)
 * Cache responses in a custom cache to be used by other services

Common use cases for transformers
 * serialize/de-serialize data, both directions
 * Strip headers (Angular 1 strips user-specified XSRF header with a default response transformation)

Worth noting are some additional features provided by Angular 1 $http, which would otherwise be handled by transformers:

 * Url-based caching
 * Configurable parameter serialization
 * Header serialization

# New Considerations for Angular 2’s Http Library

Angular 2 and Angular 2 Http have some technical differences and philosophical differences from Angular 1.
 * Mutable global state is discouraged (i.e. `$http.defaults.transforms.push(...)`).
 * Angular 2 has no "phases" like "config" and "run".
 * Angular 2 provides hierarchical dependency injection, allowing users to modify bindings at different levels of an application's component tree.
 * Angular 2 Http is based primarily on `Observable` instead of Promise
 * Http has goals of supporting upload/download progress events, connection retrying, request cancellation, caching, and polling as first-class use cases.  
 
## Current Options

The current library supports creating services that would provide shared interceptors and transformations. Here's an example of an Http-based service that would add custom headers, cache responses and log errors, based on the current RxJS implementation.

```js
import {Http, Request, IRequestOptions} from 'angular2/http';
import {ResponseCache} from './my-response-cache';
import {ErrorLog} from './my-error-log';
class MyConfiguredHttp {
	constructor(public http:Http, public responseCache:ResponseCache, public errorLog:ErrorLog) {}
	request(options:IRequestOptions) {
		if (options.url in this.responseCache){
			//Naive example (This Observable create syntax not yet supported)
			//There's a cached response for this url, return it instead of performing request
			return Rx.Observable.from([this.responseCache[options.url]]);
		}
 
		var req = new Request(options);
		// Set a custom auth header
		req.headers.set('sessionid', '123');
		return this.http.request(req)
			//Try performing the request up to three times
			.retry(3)
			.do(res => {
				//Naive example, cache the response by the original url
				this.responseCache[options.url] = res;
			})
			.doOnError(res => {
				// Log the error with my app's logging service
				this.errorLog.push(options, res);
			});
	}	
}
```

Questions to answer with final design:

 * What use cases are not able to be supported by the "shared service" approach?
 * How can the API be improved to make performance, robustness, and security the default path?
 * How is composability impacted by this approach, and how can it be improved?
 * How could this be simple and intuitive?
 * Does this require too deep of knowledge of Observables/Rx? 
 * Does anyone actually read this far down on Github issues?

**Comment by robwormald** [comment]
This sort of approach makes a lot of sense to me, and generally just feels more composable and less configure-all-the-things. All for it.

Might be worthwhile looking at doing the same sort of thing $http does, where the .then( returns the raw response, but has a couple of helper methods ( .success / .error ) which purely return the response data vs the whole response...

`http.request('foos.json').success().subscribe(....`
`http.request('foos.json').map(responseTransformer).subscribe(...`

both read nicely for me. just comes down to edumacatin' everybody on the mechanics of Observables.

**Comment by blesh** [comment]
If you wanted to do some sort of global transformers, you could have some globally registered function, but *rather than just a mapping function*, you could use a *flatMapping function* that gives you a stream of requests, and you output a stream of modified requests. i.e.:

```
Config.requestIntercept = (requests) => {
  return requests.flatMap(setHeaders);
};
```
this would allow either synchronous or asynchronous actions to occur to do things like set headers. Currently, in Angular 1, something like this would be impossible, AFAIK.

**Comment by blesh** [comment]
Further thought on this: Presuming that each `request` was an observable itself, you could actually use an Rx mechanism (the `timeout` operator) to introduce a timeout with this design, if you so chose:

```js
Config.requestIntercept = (requests) => requests.flatMap(req => req.timeout(5000))
```

**Comment by robwormald** [comment]
I'd definitely like to see more "compose / wrap it yourself" APIs vs yet-another-thing-that's-angular2-specific-that-you-have-to-look-up-in-the-docs - the current API for interceptors is a bit of a nightmare (create a factory that you register with a provider in a config block) and as @jeffbcross said, mutable global state ftl... I often talk to a couple of different APIs from an app, and global interceptors make that more complex.

**Comment by blesh** [comment]
@robwormald you're probably right. In fact, it might be better advice to tell people to compose over their requests themselves by wrapping the http service.

In fact, I think I'd like to 👎 this feature in favor of advising people to wrap the http service and use that.

**Comment by blesh** [comment]
... which is exactly what @jeffbcross had said to begin with. :P haha

**Comment by jeffbcross** [comment]
Ha, well I'd like to make sure we do due diligence, and explore ways we can make these things clear and simple, short of saying "go watch all of @blesh's RxJS tutorials, then create some wrapper services, and you're done." I tend to lean away from the declarative, "everything-on-config" approach of Angular 1, but still think there may be helpers that could be added to reduce the bar of prerequisite Rx knowledge to do common things.

**Comment by robwormald** [comment]
is it heresy to suggest we provide an `HttpClient` base class one could extend and configure themselves?

something like 
```
class MyHttpClient extends BaseHttpClient {
  transformResponse(res){  //do custom stuff here? } 
  transformRequest(req){ //do custom stuff here? }
}

myHttpClient.get('foos.json').subscribe(...)
```
I know I just said composition, and I prefer the wrap-it method, but it seems the choices are:
- extensible base class
- teach 'em how to wrap
- explaining how to do all the bind() jazz to override some injectable config.

**Comment by blesh** [comment]
Perhaps helpers could come in the favor of helpers to extend our wrap the service? I'm not sure what this looks like, just spitballing.

**Comment by lehmamic** [comment]
I would prefer a general interception mechanism in Angular2 which is not only dedicated to an http service. This would allow a lot more possibilities.

I entered an issue for that yesterday #2893

**Comment by thelgevold** [comment]
It's been a while since I looked at http in Angular 2.0. But after digging through the Angular source code, and doing a POC for my blog,  my impression was that it follows some variation on this pattern: 

http://www.syntaxsuccess.com/viewarticle/angular-2.0-and-http 

Is this pattern still part of the vision?

My initial stab at this was very simplistic and did not expose me to concepts like error handling and chaining or parallel requests. Surely there is a plan for this, but it's not 100% clear to me how it will look syntactically. Maybe not even conceptually :-)

I find the current $q based solution to be very attractive when it comes to chaining, error handling and parallel requests using $q.all(..) etc. Another big advantage is that it aligns well with the promise standard know by most developers. I know rx.js pre-dates Angular, but I am worried that it's not as widely used as the promise model and might be seen as an Angular "odd man out" approach. 

Given the huge pivot of Angular 2.0 in general I think one should be allowed to expect a very compelling argument in favor of change before making changes to things that might not even be broken. Not saying such an argument can't be produced, but at the same time, I am very interested in hearing why Angular 2.0 decided to move away from $q and $http.

I suspect there are some cool features built into rx since you decided to go in that direction. However, I think it's important to not forget that the 90% use case for http calls is most likely hitting some api and request some simple json data. My point is that there is a trade off between forcing everyone to reeducate themselves only to gain access to features that might already be solved well in the current, simple well known approach. 

That said, I am not against change, just want to learn more about the reasoning behind it.

**Comment by robwormald** [comment]
Hey Torgeir

Thanks for the input - I'll write a more in depth response later, but have a look at this plunk and it might answer some of your initial questions : http://plnkr.co/edit/ul0NVGadmGi3v6nywGRw

**Comment by thelgevold** [comment]
Thanks @robwormald I will definitely take a look at that plunk!

**Comment by e-oz** [comment]
I'm glad to see this approach, when we use DI instead of global changes. To make this approach work with 3-party libraries too, we need to encourage them to require Http as dependency (if they need Http to work).

>>> Might be worthwhile looking at doing the same sort of thing $http does, where the .then( returns the raw response, but has a couple of helper methods ( .success / .error ) which purely return the response data vs the whole response...

Please don't make this mistake again. $http.success/error is a very handy thing but it breaks the API signature - gives wrong expectation. For example, in first version of your component you return directly result of $http.get, client code written with .success/.error, and in later versions of component you find you need to return usual promise, without .success/.error functions - so you have 2 choices: breaking change or own implementation of .success/.error to your promise, just for few clients.

So please follow observables API, without additions.

**Comment by PascalPrecht** [comment]
I agree with @e-oz here to *not* introduce sth. like `.success()` and `.error()`. Those have also been deprecated in Angular 1.4 for the reasons @e-oz mentioned.

**Comment by robwormald** [comment]
The difference with the success/error idea in observable is that it's an *operator* - and adding operators is completely acceptable in Observable land (rxnext has a specific API for extending) - vs the terminator. 
- http.get(url).filter(res => res.statusCode > 200).map(res => res.json()).subscribe(onSuccess)
- http.get(url).success().subscribe(onSuccess)
- http.get(url).toJSON().subscribe(onSuccess)

Note that you're calling .subscribe() in all cases there (equiv to .then). I do take your point tho. Thanks for the input, keep it coming!

**Comment by mkretz** [comment]
@jeffbcross one use case to consider might be the following: using $http interceptors in Angular 1 you can ensure that certain behaviour (e.g. redirect for authentication) is there by default, without developers having to remember to use a  particular wrapper or shared service. Would this kind of default behaviour still be possible in Angular 2 using the wrapper approach? I am guessing you can tell DI to use the wrapper whenever Http is injected, right (do not know Angular 2 well enough yet)?

**Comment by mkretz** [comment]
@jeffbcross ... OK I think I can answer my own question now: you would use something like `provide(Http, {useClass: MyConfiguredHttp})` at the root of the component tree to override Http with the wrapper. The use case I mentioned is thus covered by the wrapper approach.

**Comment by jeffbcross** [comment]
@mkretz yep, totally fine to override the binding like that for the universal stuff like logging. Though for backend-specific logic, like redirecting to a login page on auth errors, it's probably better to create a backend-specific service to do this, and inject that service every place you want to communicate with that backend. This approach helps simplify backend-specific logic (instead of having conditional request-based logic in a single global interceptor), while also reducing side effects for third-party components in an application that make requests with the injected `Http`.

We still have a lot of work to do to document this stuff...

**Comment by angexis** [comment]
Hello,
I'll be trying to answer the questions... the one I know for sure is:
> Does anyone actually read this far down on Github issues?

Yes for sure!

Now a more controversial one:
> How could this be simple and intuitive?

I may not have taken enough time to read the RxJs docs and tutos but it seems to me that whatever Angular2 gains in simplicity wrt Angular1 is lost by the use of RxJs... Again, it's probably all my fault but that's my initial feeling when trying to deal with HTTP requests in Angular2.

> What use cases are not able to be supported by the "shared service" approach?

Since I did not (yet) manage to make my use cases work, I cannot really tell. What I can do if this is of any help is list the two use cases I have:

1. I use a HMAC authentication between my client and the server. So basically, the content of the HTTP requests is signed with a private key. For a GET, or HEAD, it is just the path of the URL which is taken into account, and for a PUT, POST, PATCH (anything that has a payload), the path+the payload are included in the digital signature.
Doing this with Angular1 was very easy and only required a single configuration to do this application-wide. Obviously, you need to be called **when the request is ready to go** and add an Authorization header to that request.

2. In order to detect any concurrent modification, all resources of the REST API are served with a hash of the delivered content (this hash is placed in a X-Etag header... it is really the same meaning as an Etag but I did not want to have issues with caching in the various proxies).
So whenever receiving a response, my Angular1 interceptor used to cache the hash in a map<resource => hash value>. So that if the client decided to modify that resource, it would issue a PUT and I would find the hash in my cache and piggyback its value in an X-If-Match header. The server would then be able to detect whether the resource was changed concurrently or not.
In that case, the "interceptor" needs to be called **as soon as the response arrives** and just before it ready to go (as in the first case).

I hope I made my use cases clear enough to be of any use... if not, please just let me know. And BTW, if you see how I could do this with the current Angular2 version... I'd be interested.

Now it's my turn to ask: does anyone actually read this far down on Github issues? ;-)

Cheers

**Comment by angexis** [comment]
Following my previous post, I struggled a bit (among others because of bug #5237) but managed to reach my goal (with alpha 47 where that bug is fixed).

Forget about my comment regarding RxJs... it's probably a very good way to go now that I learned it a bit.

However, overriding an Http service is still a quite verbose task (see my complete solution below).

One of the reasons is that all HTTP methods are implemented with the function ```httpRequest()``` defined at the top of the module, rather than using a single pipe such as ```Http.request()```. If it weren't the case, we could simple extend Http and redefine the ```Http.request()``` function, couldn't we?

Anyway, for what I am trying to do, this used to be much easier (and less copy-paste) with Angular1 than Angular2.

Any plans to make it easier before the release?

```typescript
import {} from 'base-cache/base-cache';

import {Http, Request, Headers, Response, RequestOptionsArgs, RequestOptions, RequestMethods} from 'angular2/http';
import {Observable, Injectable, Inject, forwardRef} from 'angular2/core';
import {HmacAuthenticator} from './Auth';

@Injectable()
export class MyHttpWrapper {
  private etags: Cache = new Cache({ max_size: 50 });

  constructor(private http: Http,
    @Inject(forwardRef(() => HmacAuthenticator)) private auth: HmacAuthenticator) { }

  request(url: string | Request, options?: RequestOptionsArgs): Observable<Response> {
    let req: Request;
    if (typeof url === 'string') {
      let reqOpt = new RequestOptions(options);
      reqOpt.url = url;
      req = new Request(reqOpt);
    }
    else {
      req = url;
    }

    this._beforeCall(req);

    return this.http.request(req)
      .do((res:Response) => { this._afterCall(req, res) });
  }
  get(url: string, options?: RequestOptionsArgs): Observable<Response> {
    let opts: RequestOptionsArgs = this._build(RequestMethods.Get, url, options);
    return this.request(url, opts);
  }
  post(url: string, body: string, options?: RequestOptionsArgs): Observable<Response> {
    let opts: RequestOptionsArgs = this._build(RequestMethods.Post, url, options, body);
    return this.request(url, opts);
  }
  put(url: string,  body: string, options?: RequestOptionsArgs): Observable<Response> {
    let opts: RequestOptionsArgs = this._build(RequestMethods.Put, url, options, body);
    return this.request(url, opts);
  }
  patch(url: string,  body: string, options?: RequestOptionsArgs): Observable<Response> {
    let opts: RequestOptionsArgs = this._build(RequestMethods.Patch, url, options, body);
    return this.request(url, opts);
  }
  delete(url: string, options?: RequestOptionsArgs): Observable<Response> {
    let opts: RequestOptionsArgs = this._build(RequestMethods.Delete, url, options);
    return this.request(url, opts);
  }

  private _getRequestPath(req: Request): string {
    return URI(req.url, window.location.href).path();
  }

  private _beforeCall(req: Request): void {
    let path = this._getRequestPath(req);

    // Sign the request with the user's private key
    let content = path;
    if (req.method !== RequestMethods.Get && req.method !== RequestMethods.Head) {
      content += req.text();
    }
    let authHeader = this.auth.getSignature(content);
    req.headers.set('Authorization', authHeader);

    // Add headers for concurrent changes detection
    if (req.method === RequestMethods.Put || req.method === RequestMethods.Patch) {
      let etag = this.etags.find(path);
      if (etag) {
        req.headers.set('x-if-match', etag);
      }
    }
  }

  private _afterCall(req: Request, res: Response) {
    let etag: string = res.headers.get('x-etag');
    let path = this._getRequestPath(req);
    console.log("Received: x-etag=" + etag + " from resource " + path + ": " + res.text());
    if (etag) {
      this.etags.add(path, etag);
    }
  }

  private _build(method: RequestMethods, url: string, options: RequestOptionsArgs, body?: string): RequestOptionsArgs {
    let aBody = body ? body : options && options.body ? options.body : undefined
    let opts: RequestOptionsArgs = {
      method: method,
      url: url,
      headers: options && options.headers ? options.headers : new Headers(),
      search: options && options.search ? options.search : undefined,
      body: aBody
    };
    return opts;
  }
}

```

**Comment by blesh** [comment]
Global http interceptors seem like an anti-pattern to me. They'd be better to compose in.

```js
import { interceptRequest, interceptResponse } from 'myInterceptors';

function myHttp(request: Request) {
  const newRequest: Request = interceptRequest(request);
  return http(newRequest).map(interceptResponse);
}
```

And then you'd just use `myHttp` everywhere you wanted the intercepted behavior. Ideally, you'd name it something more appropriate like "processedHttp" or "specialApiHttp" or something.

**Comment by jeffbcross** [comment]
@angexis it will usually be simpler to inject `Http` into a service than to extend it. This way, it's not necessary to manage injecting the correct dependencies into your subclass. But it's not totally unreasonable if you want to have something that behaves almost exactly like `Http` but with some different logic to communicate with a specific backend. For that use case, there's no reason we couldn't change `Http` to call `http.request` from the convenience methods, so subclasses could just override `request` and call `super.request`.

@blesh yeah, we don't want to encourage the same universally-applied global interceptors/mutable-transformers as they were in Angular 1. Creating backend-specific services that delegate to `Http`, which don't override the `Http` DI binding are legit.

**Comment by angexis** [comment]
@jeffbcross yes, that would be great if ```Http``` could call ```Http.request``` in all its methods!

Ideally, building the ```Request``` should be re-usable as well... so that only the real call to the ```ConnectionBackend``` would need to be redefined in the derived class.

Today, it is the code in the function (not a *method*)
```typescript
function httpRequest(backend: ConnectionBackend, request: Request): Observable<Response> {
  return backend.createConnection(request).response;
}
```
I am still a bit puzzled by the simpler solution you refer to... it looks already quite comfortable to me to be allowed to follow a decorator pattern (```MySpecialHttp``` extends Http **and** has the default ```Http``` implementation injected).

As I understand Angular2 DI mechanism (not 100% yet I have to confess)... I should be able to chose to use one of ```Http``` or ```MySpecialHttp``` on a case by case basis in the ```providers``` of my components, shouldn't I?

Another option I can think of is to replace the ```XHRBackend``` but that sounds more difficult to do on a local scope and I understand that *Mutable global state is discouraged* (and clearly agree with that).

Cheers

**Comment by rodrigouroz** [comment]
So, after having read all the posts and comments I'm still a little baffled about how to implement what I'm trying to achieve.

I'm trying to implement a library that I like a lot from Angular 1 (https://github.com/chieffancypants/angular-loading-bar)

In order to show in the UI that there are pending http requests, I need someway to intercept every http request to push to a stack when it starts and remove from the stack when it finishes, therefore I can show something on the screen when the stack is not empty.

So is the wrapper on the http the way to go? I would like to build an open source library so I'd like to do it following the spirit of Angular 2 of course. I'm still not sure how to do it, I guess I'll check the source code to understand it.

What do you guys think?

**Comment by jamesthurley** [comment]
I just wanted to pick up on what @thelgevold was saying about using RxJS in Angular 2's Http class.

In practical terms you can easily convert from a promise to an observable using a [single function](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/frompromise.md), and similarly you can convert from an observable to a promise with a [single function](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/topromise.md).

But given how easy the conversion is it seems surprising not to use the most standardized option by default, rather than introducing a new library that while admittedly powerful is also unfamiliar to most.

With async/await coming (and already here with Babel and TypeScript), it feels like a lot of people are going to want to do something simple like this:

```
var response = await this.http.get(uri).toPromise();
return response.json();
```

And most people will probably now start by googling how to convert form an observable to a promise (which is what I did).

I'd appreciate hearing more views on why RxJS was chosen as the default.

**Comment by thelgevold** [comment]
@jamesthurley I am still making up my mind about Promises vs Observables, but we had an interesting debate recently here: https://github.com/angular/angular/issues/5876

**Comment by blesh** [comment]
Observables can be converted to promises with `toPromise()` or `forEach()` and Promises can be converted to Observables with `Observable.from()`. Also, most operators that accept Observables accept promises.

**Comment by rodrigouroz** [comment]
I have this use case: I want to be able to do something each time an http request starts and ends.

I tried with no luck to extend the Http class and use that one.

```
import {Http, Request, Response, RequestOptionsArgs, ConnectionBackend, RequestOptions} from 'angular2/http';
import { Observable } from 'rxjs/Observable';
import {Injectable} from 'angular2/core';

@Injectable()
export class ProgressIndicatorHttp extends Http {
    
    constructor(_backend: ConnectionBackend, _defaultOptions: RequestOptions) {
        console.log('super test');
        super(_backend, _defaultOptions);
    }
    
    request(url: string | Request, options?: RequestOptionsArgs): Observable<Response> {
        console.log('test');
        return super.request(url, options);
    }
    
    get(url: string, options?: RequestOptionsArgs): Observable<Response> {
        console.log('test');
        return super.get(url, options);
    }
}
```

And then in the bootstrap:
```
bootstrap(AppComponent, [ROUTER_PROVIDERS, HTTP_PROVIDERS, provide(Http, {useClass: ProgressIndicatorHttp})]);
```

However it doesn't work. I tried several different ways, this is my latest attempt. I get: "Exception: No provider for ConnectionBackend!"

What's the correct way of doing what I'm trying to do?

The outcome is that I want to build a library that people can install to show a loading indicator when there are pending http requests.

Thanks and Merry Christmas to all who observe it!

**Comment by angexis** [comment]
@rodrigouroz you can achieve that with the sample code I posted above...

**Comment by rodrigouroz** [comment]
@angexis so you can't just extend the class and add your logic and then calling the parent method?
It's very error prone to me to just have to reimplement the Http class yourself. Of course it's a copy and paste but if later on the implementation changes you end up with an old implementation. The extension would be much clearer IMHO.

Thanks, I'll give it a try.

**Comment by riaann** [comment]
Another use-case for the list: being able to transparently batch requests occurring within a short time of each other to reduce round trips (until HTTP/2 is more ubiquitous). This is especially useful for ReST APIs where page navigation could trigger multiple requests.  

This is an example of a library that implements this strategy for Angular 1: jonsamwell/angular-http-batcher

**Comment by jamesthurley** [comment]
@rodrigouroz you don't have to re-implement it as such, it's just using composition rather than inheritance.

A more basic example that you could build off would be:

```
export class CustomHttp {
  constructor(private http: Http) {
  }

  public request(url: string | Request, options?: RequestOptionsArgs) {
    return this.http.request(url, options);
  }

  public get(url: string, options?: RequestOptionsArgs) {
    return this.http.get(url, options);
  }

  public post(url: string, body: string, options?: RequestOptionsArgs) {
    return this.http.post(url, body, options);
  }

  public put(url: string, body: string, options?: RequestOptionsArgs) {
    return this.http.put(url, body, options);
  }

  public delete(url: string, options?: RequestOptionsArgs) {
    return this.http.delete(url, options);
  }

  public patch(url: string, body: string, options?: RequestOptionsArgs) {
    return this.http.patch(url, body, options);
  }

  public head(url: string, options?: RequestOptionsArgs) {
    return this.http.head(url, options);
  }
}

```

**Comment by rodrigouroz** [comment]
@jamesthurley 

Thanks, now I got the concept and my example works.

However I realized that if I write a library I need to instruct people to replace Http all over their codebase with the new class, because when I bootstrapped the HTTP_PROVIDERS and put provide(Http, {useClass: myClass}) I got a circular DI (since myClass uses Http, makes sense). I couldn't find a way to do it automatically.

Basically I wanted a way to say: "each time you require Http, use this class instead". I'm trying to build a library that will show a loading indicator when there are pending network requests and I'm trying to make the installation easy.

Thank you all for your help.

Rod

**Comment by rodrigouroz** [comment]
Now I think what I need is to decorate the ConnectionBackend, since I want only to know when a request starts and end, I don't care about the contents...

**Comment by angexis** [comment]
@rodrigouroz yes it is very error prone to code all this again!

This is the reason I was asking for a change (this is an RFC after all) in the base Http implementation (to allow for more re-use)... Whether we inherit, delegate or both (decorator), is a secondary topic.

Good luck anyway!

**Comment by jamesthurley** [comment]
Perhaps @rodrigouroz is an example of where a built in (but not global) interceptor system would be beneficial.

For example, if there were an `InterceptingHttp` class that consumed `Http` but added calls to registered interceptors, then i could create `MyApiHttp` that looked something like this:

```
export class MyApiHttp {
  constructor(
    private http: InterceptingHttp, 
    progress: ProgressInterceptor, 
    batcher: BatchingInterceptor) {

    this.http.registerInterceptors(progress, batcher);
  }

  public request(url: string | Request, options?: RequestOptionsArgs) {
    return this.http.request(url, options);
  }

  public get(url: string, options?: RequestOptionsArgs) {
    return this.http.get(url, options);
  }

  // etc...
}
```

This way I can still have different interceptor configurations for different APIs, but people like @rodrigouroz have a simple, standard way of providing generic interceptor services.

Of course none of that needs to be done as part of the core Angular 2.

**Comment by rodrigouroz** [comment]
I finally solved my case by injecting a custom XHRBackend like this:

```
bootstrap(AppComponent, [HTTP_PROVIDERS, provide(XHRBackend, {useClass: ProgressIndicatorBackend}));
```

And the class `ProgressIndicatorBackend`

```
import {ConnectionBackend, Connection, Request, Response, ReadyState, XHRConnection, BrowserXhr, ResponseOptions} from 'angular2/http';
import {Injectable} from 'angular2/core';
import {Observable} from 'rxJs/Observable';

class ProgressIndicatorConnection implements Connection {
    private baseConnection: XHRConnection;
    private static pendingRequests: number = 0;
    
    constructor(req: Request, browserXHR: BrowserXhr, baseResponseOptions?: ResponseOptions) {
        this.baseConnection = new XHRConnection(req, browserXHR, baseResponseOptions);
        ProgressIndicatorConnection.pendingRequests++;
        console.log('Pending requests: ' + ProgressIndicatorConnection.pendingRequests);
        this.response.subscribe(() => {
           ProgressIndicatorConnection.pendingRequests--; 
           console.log('Pending requests: ' + ProgressIndicatorConnection.pendingRequests);
        });
    }
    
    get readyState(): ReadyState {
        return this.baseConnection.readyState;
    }
    get request(): Request {
        return this.baseConnection.request;
    }
    get response(): Observable<Response> {
        return this.baseConnection.response;
    }
    
}

@Injectable()
export class ProgressIndicatorBackend implements ConnectionBackend {
  constructor(private _browserXHR: BrowserXhr, private _baseResponseOptions: ResponseOptions) {}
  createConnection(request: Request): ProgressIndicatorConnection {
    return new ProgressIndicatorConnection(request, this._browserXHR, this._baseResponseOptions);
  }
}
```

Now my next step is to understand how to expose the number of pending request for another component to subscribe to it (and turn the loading indicator on and off). How can I achieve that? In Angular 1 I would use the $rootScope to broadcast an event...

**Comment by jamesthurley** [comment]
@rodrigouroz you can use an [EventEmitter](https://angular.io/docs/ts/latest/api/core/EventEmitter-class.html) to expose an event from your service.

**Comment by rodrigouroz** [comment]
@jamesthurley I was just reading that page. I'm trying to understand how to subscribe to the event from an external component. The emitter will be the ProgressIndicatorConnection class I just copied above.

**Comment by jamesthurley** [comment]
@rodrigouroz you can inject the ProgressIndicatorConnection into your other component and call the `subscribe` method, passing in a callback.  I don't think this is a discussion for this thread however...

**Comment by rodrigouroz** [comment]
I can't find a way of using the EventEmitter, since I don't have a Component.

instead I'm thinking of Injecting the ProgressIndicatorConnection in the component that shows the loading indicator and create an observable. That's the only way I can think of...

**Comment by rodrigouroz** [comment]
This is the final solution for anyone interested. I'm open to suggestions and comments of course.

```
import {ConnectionBackend, Connection, Request, Response, ReadyState, XHRConnection, BrowserXhr, ResponseOptions} from 'angular2/http';
import {Injectable} from 'angular2/core';
import {Observable} from 'rxjs/Observable';

export class ProgressIndicatorConnection implements Connection {
    private baseConnection: XHRConnection;
    private static pendingRequests: number = 0;
    private static observer;
    static pending: Observable<boolean> = new Observable (observer => ProgressIndicatorConnection.observer = observer);
    
    constructor(req: Request, browserXHR: BrowserXhr, baseResponseOptions?: ResponseOptions) {
        this.baseConnection = new XHRConnection(req, browserXHR, baseResponseOptions);
        ProgressIndicatorConnection.requestStarted();
        this.response.subscribe(() => {
           ProgressIndicatorConnection.requestEnded(); 
        });
    }
    
    static requestStarted() {
        if (ProgressIndicatorConnection.pendingRequests == 0) {
            ProgressIndicatorConnection.observer.next(true);
        }
        ProgressIndicatorConnection.pendingRequests++;
    }
    
    static requestEnded() {
        if (ProgressIndicatorConnection.pendingRequests == 1) {
            ProgressIndicatorConnection.observer.next(false);
        }
        ProgressIndicatorConnection.pendingRequests--;
    }
    
    get readyState(): ReadyState {
        return this.baseConnection.readyState;
    }
    get request(): Request {
        return this.baseConnection.request;
    }
    get response(): Observable<Response> {
        return this.baseConnection.response;
    }
    
}

@Injectable()
export class ProgressIndicatorBackend implements ConnectionBackend {
  constructor(private _browserXHR: BrowserXhr, private _baseResponseOptions: ResponseOptions) {}
  createConnection(request: Request): ProgressIndicatorConnection {
    return new ProgressIndicatorConnection(request, this._browserXHR, this._baseResponseOptions);
  }
}
```

**Comment by robwormald** [comment]
@angexis see https://github.com/angular/angular/pull/6198 - should give you the ability to subclass and override request when it lands.

**Comment by angexis** [comment]
@robwormald looks pretty cool! Thanks!

**Comment by stefanocke** [comment]
For those who struggled with circular DI when wrapping the Http service: The following did the job for me. Just use MY_HTTP_PROVIDERS in your bootstrap and all your services and components will get the wrapped Http injected.


```
export const MY_HTTP_PROVIDERS = [
    HTTP_PROVIDERS,
    provide(Http,
        {
            useFactory: (xhrBackend: XHRBackend, requestOptions: RequestOptions) => {
                let originalHttp = new Http(xhrBackend, requestOptions);
                return new MyHttpWrapperService(originalHttp);
            },
            deps: [XHRBackend, RequestOptions]
        })];
```

**Comment by forfuns** [comment]
@rodrigouroz  , hi,I have some demand like you did,but I try your code , is no work in my app, like the `ProgressIndicatorConnection` and `ProgressIndicatorBackend` both not have instantiationI

**Comment by lordgreg** [comment]
Could someone provide a basic app example with an http-interceptor which, doesn't implements angular2 `Http` class? I'd love to do the interceptor, one problem being exactly what @stefanocke wrote, I'm getting Http > Http circular dependencies, where my HttpInterceptor injects Http.

app.ts
```
import {provide} from 'angular2/core';
import {Http} from 'angular2/http';
import {HttpInterceptor} from './providers/http-interceptor/http-interceptor';

@App({
  templateUrl: 'build/app.html',
  prodMode: true,
  config: {},
  providers: [
    provide(Http, { useClass: HttpInterceptor })
  ]
})
export class MyApp {

  constructor(
    private http: Http
  ) {
```

and Interceptor
```
import {Injectable} from 'angular2/core';
import {Http, Connection, Headers, Request, RequestOptions, RequestMethod, RequestOptionsArgs} from 'angular2/http';

@Injectable()
export class HttpInterceptor {
  constructor(private http: Http) { ... }
```

Somehow I don't know how to include the `MY_HTTP_PROVIDERS` in apps providers part and then import Http class as I did until now in all other @Pages and @Components

**Comment by zoechi** [comment]
@lordgreg In your application just ignore `HttpInterceptor`. If you register `MY_HTTP_PROVIDERS` in `bootstrap(...)` your classes will always get `HttpIntercepter` when they request `Http`.

From `app.tx` just remove

    import {HttpInterceptor} from './providers/http-interceptor/http-interceptor';

and

    providers: [
      provide(Http, { useClass: HttpInterceptor })
    ]

**Comment by byavv** [comment]
@lordgreg, I have one, take a look at this: [using injected Http](https://github.com/byavv/mea2n/blob/master/src/client/app/common/services/authHttp.service.ts)
and this: [extending http](https://github.com/byavv/ng2-playground/blob/master/client/app/modules/http_explore/extHttp.ts). I prefer first, but you can compare.

**Comment by flofly** [comment]
@lordgreg thanks for you code sample, very helpful. I will just add the logic to use a refresh token.

**Comment by flofly** [comment]
@byavv thanks for you code sample, very helpful. I will just add the logic to use a refresh token.

**Comment by byavv** [comment]
@flofly,  glad I could help

**Comment by bennadel** [comment]
When I first read this issue, I was disgruntled over the fast that *I* had to implement logic on my own (especially logic that was already implemented in AngularJS 1.x). But, after more reflection and some experimentation, I have to admit, I really like it. When I really sat down and thought about it, basically none of my "global configurations" actually made sense at a _global level_. It really does make more sense to have all specifics pushed into a specialized API client. 

Anyway, just wanted to share perspective from someone who thought this was initially bad, but then came to see the light of the decision. I tried to [write up some of my thoughts here](http://www.bennadel.com/blog/3047-creating-specialized-http-clients-in-angular-2-beta-8.htm) for an HTTP client that implements:

* Encapsulation of the HTTP transportation mechanism (including Content-Type headers).
* Param and Data driven URL interpolation.
* Automatic JSON serialization for outgoing requests.
* Automatic JSON parsing for incoming responses.
* Normalized error responses.
* XSRF (Cross-Site Request Forgery) protection.
* Pending request tracking.
* Explicit handling of certain HTTP response codes (ex, 401 Unauthorized).

**Comment by robwormald** [comment]
Great write up, thanks @bennadel!

**Comment by danicomas** [comment]
Nice @bennadel

**Comment by bennadel** [comment]
@robwormald @danicomas much appreciated, fine gentlemen.

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166448.
------------------------------------------------------------------------------------------------------------


[[0ea36354c24bd4069841dfb5f16d146ea1c3e9d5-673073314]]

**Comment by MarkPieszak** [comment]
Nice article Ben! How come you're not using TypeScript? Just curious! @bennadel

**Comment by bennadel** [comment]
@MarkPieszak thank you, good sir. I just happen to really enjoy ES5. That, and I can't figure out yet how to use TypeScript AND keep all my code on one page for my demos. If I could figure that out, I might start using TypeScript more.

**Comment by witoldsz** [comment]
Hi guys,
This request became a thread of so many comments that it's easy to get lost and I am not sure I followed everything.

It looks like the solution mentioned here the most is to create a custom Http service or to use the default one with custom dependencies. I really have bad feeling about that approach. Let's omit the fact that this is complicated and focus on something else: composition. If each of us have created our own, custom solution, how could we share it with each other and compose?

@bennadel implemented some ApiGateway service with like... 8 nice and shiny features. One day he finds something similar with 6 super features and he finds 3 of them attractive. Since these two implementations are monolithic how to combine them? What would you do, @bennadel if you were to create another, smaller application, and you would like to "borrow" 2 of 8 features of ApiGateway?

Years ago I have described and then implemented an Angular HTTP Auth interceptor. I have found it really useful in other applications of mine, so I have extracted it and published on GitHub. I was able to keep it separated and maintain it outside of my apps. I could use it with other HTTP interceptors and other developers could easily use AND combine that with other interceptors. 

How is that supposed to happen in Angular2 if everyone has to create their own version of custom ApiGateway-like? @jeffbcross request is all about a common API, so all of us could tinker with Http with simple, exchangeable and composable features.

I don't know much about Angular 2 yet but it looks like simple function composition for Http, configured by injection API would do the job just fine.

**Comment by keithdmoore** [comment]
In case someone finds this issue during a Google sleuthing session.  I found this issue to be helpful when trying to implement a loading/progress spinner.  https://github.com/angular/http/issues/80  I ended doing something similar to this: https://plnkr.co/edit/VVOGaoZjcWZqTjCUJ6HC?p=preview which is mentioned in that issue as well.

**Comment by squadwuschel** [comment]
@byavv I tried to use the second example, but

   let requestOptions = new RequestOptions(Object.assign( ...

there is no assign method on Object. And could you please explain or extend your example how its working with a real service and what I have to register in boot and in my service (I am a bit new to ng2)?

@jeffbcross I found the old solution for Interceptors in NG1 very good and simple. I've used it a lot for showing global messages - I've just looked up the request object for the default object where my error or success messages are included and I've extracted the messages and added them to the messageService. Or when I got a server error 500 I used the Interceptor to redirect to my error view and I only have to do it once such a feature would be awesome in NG2. When I've read the above posts right, the current implementation for such a feature is not so easy to implement and understand. A Interceptor would be something that I would expect to be implemented in the ng2 framework.

**Comment by nolazybits** [comment]
@witoldsz have you found any solution
@all anyone keen to answer @witoldsz comments?
I am really keen to know how this can be solved.

Thanks

**Comment by tomascharad** [comment]
Hi guys,

I don't know if this is the correct place to ask this, but does anyone knows how to intercept a redirect?
My app uses CAS as authentication and I need to navigate to the browser to the location redirected. The current behaviour is that http goes to the redirect and tries to retrieve a response, the problem with this is that the CAS server doesn't allows CORS so the request is rejected.

**Comment by gund** [comment]
Hi everyone,  
  
I've recently started to build Http Interceptor service which will expose API for registering custom interceptor all across application (or not - may be customisable).  
  
Whole process takes 2 stages:
1. Extend Http with private API and provide provider for replacing default Http  
2. Create HttpInterceptor Service which will provide public API to manage interceptors and internally utilise private API of extended Http  
(Here probably it will use newly created http class instead of default one and user will be able to choose if he wants to replace default Http with extended one or keep it and still be able to use extended http service separately)  
  
With that said I already built POC and published key concepts of internal design and public API here:  
https://gist.github.com/gund/1072056dd3af86f084ae3a111946415b
  
If you have some ideas how I can improve it in terms of public API or internal design - comment's are much appreciated.  
  
Thanks.

**Comment by fkolar** [comment]
Just a question why this issue is closed when there is no resolution to it and I see state  -> needs a Design?

Does it mean that its not on anyones priority list you might get back to in some future angular's version?

**Comment by JanStureNielsen** [comment]
@fkolar -- it looks like @robwormald replaced this issue with [issue #80 in the http module](https://github.com/angular/http/issues/80).

**Comment by fkolar** [comment]
@JanStureNielsen => are you sure its #80 ?

**Comment by MarkPieszak** [comment]
Issue 80 in the http repo: https://github.com/angular/http/issues/80 @JanStureNielsen

**Comment by voliva** [comment]
angular/http repo is down? Where can we find that issue now?

I've made a npm package here that provides a Http-like service with interceptor registering functionalities: https://github.com/voliva/angular2-interceptors

**Comment by piernik** [comment]
I'm also looking for easy, angular1 like, solution.

**Comment by squadwuschel** [comment]
is there a out of the box solution for Interceptors on the roadmap for ng2 or do I have to solve the problem?

**Comment by DzmitryShylovich** [comment]
you have to solve it yourself

**Comment by witoldsz** [comment]
The problem with DIY solutions is that they are not going to be compatible with each other…

**Comment by kentoj** [comment]
Here is an excellent conversion of Ben Nadel's custom HTTP client.
https://blog.sstorie.com/adapting-ben-nadels-apigateway-to-pure-typescript/

@witoldsz It seems like DIY solutions are not supposed to be compatible with each other for the most part because they are specific to the project. For example, security config for headers and tokens will differ across projects. However, there are somethings that could be common, such as some of the functions in the example I linked like URL path variable parsing, adding content type, and adding a xsrf cookie.

**Comment by voliva** [comment]
@kentoj I tried creating a npm package to provide an interface for registering individual interceptors https://github.com/voliva/angular2-interceptors. The main idea is to have a structure that would support this kind of use case, and then you'd have interceptors specific for each project, although many of them could be shared.

I'm looking forward to improve my solution, so any kind of feedback is welcome :)

**Comment by LiorArbel** [comment]
@MarkPieszak Hey Mark just wanted to bring to your attention that  [issue #80](https://github.com/angular/http/issues/80) doesn't exists, might be mainly because the angular/http repository doesn't exist (or is private?) currently.

I beg you, please provide some kind of a simple solution and\or check what is going on with the http repository... I really really don't want to rewrite so much of the http class for such simple functionality... what do we have the DRY rule for??

**Comment by MarkPieszak** [comment]
It went private then I believe was merged into the main [Angular repo](https://github.com/angular/angular/tree/master/modules/%40angular/http).

I believe (but I'm not sure) that at least the current goal is to leave it up to the user to override or implement Http and wrap things that way.  But I do agree, it would be nice to keep things consistent between ng1&2 in this regard.

What are you currently doing to achieve this?  There's a lot of options online to choose from as well.

**Comment by kentoj** [comment]
@LiorArbel and @MarkPieszak At first glance it definitely seemed like a violation of DRY until I did my own implementation of an HTTP client. What it ended up being was a service for accessing my backend API gateway. 

Now that my first implementation is done I could see a valid use case for creating a custom HTTP client for every API gateway my application interacts with. Why? the authentication may be different, the content type may be different (XML or JSON), one may implement hypermedia (HATEOAS) and one may not, and so forth.

Creating a global HTTP interceptor could make things harder for me if I interact with multiple APIs that have different requirements. This could be the 80/20 rule though. If 80% of angular 2 apps only communicate to one backend API then a global interceptor may not be so bad.

The HTTP interceptor module, if created, seems like a module that could be added in and used conditionally, similar to how Spring Java uses the `@Conditional` annotation to conditionally wire up and configure components based on whether they are present.

At the very least we should point from the official docs to an example or at least the idea of a custom HTTP client, the one I mentioned from Sam Storie earlier being a good candidate.

**Comment by witoldsz** [comment]
@kentoj this is so much the "my point of view is better than yours", actually. You have some special setup for which the "global" HTTP interceptor would be less optimal than your own custom HTTP service wrapper. That is fine, but having an option to register/plug-into the official HTTP service would suit others just well, would allow for a 3rd party market of extensions (like it's now for AngularJS 1), and you could still choose your own wrapper solution.

**Comment by kentoj** [comment]
@witoldsz How do you propose to implement the option of registering an interceptor or plugging it in? Since, as the OP stated, *Angular 2 has no "phases" like "config" and "run"*  then as Angular currently exists you would have to ensure the interceptor was registered before any HTTP calls were made. 

Or, you could do something like what @voliva did that ends up with injection of an interceptor-aware wrapper for HTTP:
```
    constructor(
        private http: InterceptorService) {
    }
```

Without adding something like "config" or "run" it looks like the best option for interceptors in Angular 2 HTTP is a wrapper, whether that be a custom HTTP client or an interceptor-aware extension to HTTP, which is a custom HTTP client. So either way, it looks like a custom HTTP client of some sorts is the way to go unless you propose adding some config phase to the angular 2 bootstrapper or dependency injector.

**Comment by LiorArbel** [comment]
@kentoj let me be a bit rude here... 
I don't see how it remotely matters that there are so many niche use cases for an http class - why would the addition of an **optional ability** to add an interceptor even matter to you??? 

On the contrary, it would only help all those tons of people that wrapped the Http class and wrote tons of coupled code to add such a simple feature from angular 1 that involves about 10 lines of code. 

I mean, even auth0 ended up (rewriting all the Http class' function signatures)[https://github.com/auth0/angular2-jwt/blob/master/angular2-jwt.ts]
I can only imagine how many angular 2 projects would get invalid if one of these functions would change.

**Comment by kentoj** [comment]
@LiorArbel I appreciate the energy. You make a good point on coupling to the Http class's API. 

I make no argument against adding an interceptor. Furthermore, I would probably use the interceptor functionality if it existed. I must be coming across as against interceptors. I am not against them at all. They were the first thing that I looked for beyond simple HTTP in an angular app. When I first learned I would have to subclass HTTP to add my functionality I was very displeased because of the added complexity, coupling, and class inheritance. Avoid class-based inheritance except when I know for sure it is the best tool for the job seems to keep me out of inheritance hierarchy troubles.

Regarding custom HTTP clients I only acknowledge the usefulness of a custom HTTP client and the value I got from having to write one myself (copy from Sam Storie and Ben Nadel :] ). Once I wrote my custom HTTP client I felt like I had more control and understanding of the underlying process. It feels like I can contain the leaks of the  network and HTTP abstractions in one place instead of everywhere I need to make an HTTP call. 

Additionally, when I implement [HATEOAS](https://spring.io/understanding/HATEOAS) since HATEOAS is a protocol for defining domain-specific languages in the future I see myself handling that protocol in a custom HTTP client and not an interceptor since the scope for that protocol management goes beyond simple interception. Since some of my backend microservices may not yet to HATEOAS I may need a custom HTTP client for each microservice depending on its maturity level. 

Even still, I can see myself using an HTTP interceptor for all requests for authentication since I expect authentication to be globally the same for one client application.

Comparing the coupling of the custom HTTP client to an interceptor API makes it seem like using an interceptor would only be coupled to one method, the interceptor registration call. 

And yet, the claim I previously made still stands. How would we add the interceptor functionality to Angular 2 without adding bootstrap or config phases? Bootstrap and config phases could be the right answer. Spring does a similar thing when doing dependency injection for a Spring Java app. Maybe the right place to add bootstrap or config for Angular 2 is in the dependency injector. That may or may not be an option for Angular 2. Who is the creator of the injector?

**Comment by ziv** [comment]
Although I found HTTP interceptors very useful in Angular 1,  there are also some drawbacks for using global interceptors. For example, I used interceptors to manage the "login" status of the user, but! what happened when you're dealing with the login request itself? Should it "trap" using the global interceptor?

I found that for complicated apps, there are several types of HTTP connections I want to handle and a global one break this approach. So, instead of manipulating the Http class itself, I have several "backend" classes that handle HTTP requests, each one responsible for its specific task (checking for login during request, handling JSON, etc.)

Changing the HTTP client itself, limit your application to create different request types.

**Comment by DzmitryShylovich** [comment]
They have ideas to introduce something like `HttpService` based on fetch api (when they add abort https://github.com/whatwg/fetch/issues/27). With async json parsing, support for interceptors etc. But there're some problems with `fetch` like it doesn't support progress at all https://github.com/github/fetch/issues/89.

**Comment by LiorArbel** [comment]
@DzmitryShylovich Thank you very much, it's nice to know. Would be happy if you could keep us updated.

@kentoj Must we have an initialization\bootstrap\config\w.e. stage? What if the Http service would be a stateful service that holds interceptors that can be added at runtime? 

And then if we take for example the authentication token header case - only after the usser logged to your server you will then push the interceptor which handles the authentication token headers.
(I have never read a single line of the angular 2 implementation so I have no iead if that is actually feasible, hope I'm not embarrassing myself (: )

**Comment by gund** [comment]
@LiorArbel good point about coupling with Http methods!
That is why I used Proxy to not to worry about Http implementation at all while building [ng-http-interceptor](https://github.com/gund/ng-http-interceptor) =)

**Comment by DzmitryShylovich** [comment]
Interceptors will be available in ~4.1 
https://docs.google.com/document/d/1C_wzuuQEjIktJxrTlUTjTCDEZR8YduGTfbpe4WsNfNk/edit#heading=h.xgjl2srtytjt

**Comment by e-cloud** [comment]
@DzmitryShylovich It seems the proposal is obsolete.

What's the status of it?
Is there a new proposal?
Or just no interceptor feature in the near future?

**Comment by bgever** [comment]
@jeffbcross Could you please update on the status?

The regression vs Angular 1.x of not having interceptors and transformers is breaking **essential** functionality we need to:
1. Make even a *simple* API call with an Authorization header.
2. Logging errors.
3. Handling retries for flaky connections (China).

To be frank, I'm stunned that the framework still does not support this and that Http is marked experimental. This has been solved years ago in v1, and XHRs are so fundamental to a SPA.

Currently the **status of this feature is in limbo**:
1. This issue is marked Closed.
2. The [backlog](https://github.com/angular/angular/issues/2793) is Closed and outdated.
3. The [referenced design doc](https://docs.google.com/document/d/1C_wzuuQEjIktJxrTlUTjTCDEZR8YduGTfbpe4WsNfNk/) is marked obsolete.
4. There is [no open tagged issue](https://github.com/angular/angular/labels/comp%3A%20http) related to this.

I don't know whether we are forced to make our own Http wrapper with this basic functionality and have to update all our dependencies. And then, when (if?) v4.1 (or a later version) supports it, we need to refactor our codebase again.

At least having clarity in this matter would soften the hurt a bit, but this migration is starting to look really ugly. 😞

**Comment by zoechi** [comment]
@bgever you can just create a custom `Http` class that extends Angulars `Http` class and customize there to your liking. There are several answers on SO how to do that.

**Comment by BorntraegerMarc** [comment]
@zoechi I think this discussion is more about how not to build a custom class but rather really only use Http interceptors. Which are two different use cases in my opinion.

**Comment by gund** [comment]
Whoever needs an elegant solution for interceptors, you can try [ng-http-interceptors](https://github.com/gund/ng-http-interceptor) =)

**Comment by bgever** [comment]
@BorntraegerMarc - spot on!
@gund - great library, thanks for the reference!

Considering not an exact port of AngularJS to Angular 4+, the custom `Http` classes do provide a more explicit way describe in the code that HTTP transformations are needed. E.g. we could have 2 custom classes for two different cross-origin APIs, having their own bearer tokens applied. It removes the "magic" of the interceptors. Also, I now understand the [Observables allow a way to retry and cancel](https://github.com/angular/angular/issues/5876), neat! It's growing on me. 😄

**Comment by imgx64** [comment]
According to [this](https://github.com/angular/angular/pull/6198#issuecomment-296929974):
> Similar stuff is planned for the new HttpClient

and

> I don't think there is a public design doc yet

So I guess something is being cooked internally at Google and they won't tell us about it until it's ready. ¯\\\_(ツ)\_/¯

**Comment by aitboudad** [comment]
see https://github.com/angular/angular/pull/17143
